### DISCUSSION TITLE
Declare an InjectionPointInfo manually

### DISCUSSION THREAD

**Discussion by scrocquesel** [quarkusio/quarkus#38317] [Q&A] [answered]
I have an extension that produce a synthetic bean only when an `InjectionPointInfo` in the `BeanRegistrationPhaseBuildItem` matches the produced bean type.
Now, I have another extension that would do the same thing for another bean and use the `SyntheticBeanBuildItem.addInjectionPoint` to inject the first bean.

Unfortunately, the first bean is not produced because the synthetic injection point is not discovered. Would it be possible to declare this InjectionPoint in a previous build step so the first bean can be produced with the same logic ?

**Comment by scrocquesel** [comment]
I split the discovery and the requirement with an intermediate build item that other extensions can contribute to.

Still, if there is a shorter way of doing it, i'm still interested.

**Comment by mkouba** [comment, marked as answer]
> Would it be possible to declare this InjectionPoint in a previous build step so the first bean can be produced with the same logic ?

Unfortunately, you cannot define a synthetic injection point beforehand. Furthermore, the set of injection points accessible through the `BeanDiscoveryFinishedBuildItem#getInjectionPoints()` or `SynthesisFinishedBuildItem#getInjectionPoints()` is updated with synthetic injection points after all synthetic beans are registered.

> I split the discovery and the requirement with an intermediate build item that other extensions can contribute to.

I'm not quite sure I understand this solution. Do I understand it correctly that the first extension produces a specific build item which signals that a synthetic bean was produced? In other words, a specific build item replaced the injection point check in another extension. If so then it's ok and I'm not aware of a better way.
